be/dtm0_log.c

```diff
@@ -0,0 +1,631 @@
+					      rec->dlr_txd.dtd_ps.dtp_nr);
+	}
+
+	return 0;
+}
+
+M0_INTERNAL int m0_be_dtm0_log_update(struct m0_be_dtm0_log  *log,
+                                      struct m0_be_tx        *tx,
+                                      struct m0_dtm0_tx_desc *txd,
+                                      struct m0_buf          *payload)
+{
+	struct m0_dtm0_log_rec	*rec;
+
+	M0_PRE(payload != NULL);
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tx_desc__invariant(txd));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	return (rec = m0_be_dtm0_log_find(log, &txd->dtd_id)) ?
+                be_dtm0_log__set(log, tx, txd, payload, rec) :
+                m0_be_dtm0_log__insert(log, tx, txd, payload);
+}
+
+M0_INTERNAL int m0_be_dtm0_log_prune(struct m0_be_dtm0_log    *log,
+                                     struct m0_be_tx          *tx,
+                                     const struct m0_dtm0_tid *id)
+{
+	/* This assignment is meaningful as it covers the empty log case */
+	int                     rc = M0_DTS_LT;
+	struct m0_dtm0_log_rec *rec;
+	struct m0_dtm0_log_rec *currec;
+
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tid__invariant(id));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	m0_tl_for (lrec, log->dl_tlist, rec) {
+		if (!m0_dtm0_tx_desc_state_eq(&rec->dlr_txd,
+					      M0_DTPS_PERSISTENT))
+			return M0_ERR(-EPROTO);
+
+		rc = m0_dtm0_tid_cmp(log->dl_cs, &rec->dlr_txd.dtd_id, id);
+		if (rc != M0_DTS_LT)
+			break;
+	} m0_tl_endfor;
+
+	if (rc != M0_DTS_EQ)
+		return -ENOENT;
+
+	while ((currec = lrec_tlist_pop(log->dl_tlist)) != rec) {
+		M0_ASSERT(m0_dtm0_log_rec__invariant(currec));
+		be_dtm0_log_rec_fini(&currec, tx);
+	}
+
+	be_dtm0_log_rec_fini(&currec, tx);
+	return rc;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_clear(struct m0_be_dtm0_log *log)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	/* This function is expected to be called only on the client side where
+	   the log will always be a volatile log. */
+	M0_ASSERT(!log->dl_is_persistent);
+
+	m0_tl_teardown(lrec, log->dl_tlist, rec) {
+		M0_ASSERT(m0_dtm0_log_rec__invariant(rec));
+		M0_ASSERT(m0_dtm0_tx_desc_state_eq(&rec->dlr_dtx.dd_txd,
+						   M0_DTPS_PERSISTENT));
+		be_dtm0_log_rec_fini(&rec, NULL);
+	}
+	M0_POST(lrec_tlist_is_empty(log->dl_tlist));
+}
+
+M0_INTERNAL int m0_be_dtm0_volatile_log_insert(struct m0_be_dtm0_log  *log,
+					       struct m0_dtm0_log_rec *rec)
+{
+	int rc;
+
+	/* TODO: dissolve dlr_txd and remove this code */
+	rc = m0_dtm0_tx_desc_copy(&rec->dlr_dtx.dd_txd, &rec->dlr_txd);
+	if (rc != 0)
+		return rc;
+
+	lrec_tlink_init_at_tail(rec, log->dl_tlist);
+	return 0;
+}
+
+M0_INTERNAL void m0_be_dtm0_volatile_log_update(struct m0_be_dtm0_log  *log,
+						struct m0_dtm0_log_rec *rec)
+{
+	/* TODO: dissolve dlr_txd and remove this code */
+	m0_dtm0_tx_desc_apply(&rec->dlr_txd, &rec->dlr_dtx.dd_txd);
+}
+
+M0_INTERNAL int m0_be_dtm0_plog_can_prune(struct m0_be_dtm0_log	    *log,
```

Comment:
Please add comments for this function.
What is the return value?
It should be bool?
In C programming language,
a non-zero (positive or negative) value means true.
But here in this function, it seems "zero" is true. 
Please modify this.

Reply:
I agree the name is misleading but the function itself isn't implemented as a bool. It is returning integer return values including error codes and also fills the input credit variable. However, now that you mention, it should indeed be implemented as returning a bool.
I have updated the function code to return true/false (besides filling the credit structure) instead of actual error codes. Thanks.